File: bindings/java/src/main/java/com/agentharness/adapters/FileAudit.java
package com.agentharness.adapters;

import com.agentharness.ports.AuditPort;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;
import java.util.function.UnaryOperator;

/**
 * Durable, append-only file AuditPort (harness-protocol.md §7.3).
 *
 * <p>Writes one JSON object per line (JSONL) to a file in append mode. There is <b>no</b> update or
 * delete API — the log is append-only by construction (INV-4). PII is redacted before every write.
 * Dependency-free (a small hand-rolled JSON writer); a JDBC / object-store adapter follows the same
 * contract.
 */
public final class FileAudit implements AuditPort {

    private static final String DECISION = "decision";
    private static final String SECURITY = "security_event";

    private final Path path;
    private final UnaryOperator<String> redact;
    private final Object lock = new Object();

    public FileAudit(Path path) {
        this(path, Redaction::redact);
    }

    public FileAudit(Path path, UnaryOperator<String> redact) {
        this.path = path;
        this.redact = redact;
        try {
            if (path.getParent() != null) {
                Files.createDirectories(path.getParent());
            }
        } catch (IOException e) {
            throw new UncheckedIOException("cannot create audit directory", e);
        }
    }

    @Override
    public void record(AuditEntry e) {
        StringBuilder sb = new StringBuilder("{");
        field(sb, "type", DECISION).append(',');
        field(sb, "agent_name", e.agentName()).append(',');
        field(sb, "tenant_id", e.tenantId()).append(',');
        field(sb, "action", e.action()).append(',');
        raw(sb, "confidence", Double.toString(e.confidence())).append(',');
        raw(sb, "auto_enforced", Boolean.toString(e.autoEnforced())).append(',');
        field(sb, "rationale", redact.apply(e.rationale())).append(',');
        field(sb, "outcome", e.outcome()).append(',');
        nullableField(sb, "correlation_id", e.correlationId()).append(',');
        field(sb, "recorded_at", e.recordedAt().toString());
        sb.append('}');
        append(sb.toString());
    }

    @Override
    public void recordSecurityEvent(SecurityEvent e) {
        StringBuilder sb = new StringBuilder("{");
        field(sb, "type", SECURITY).append(',');
        field(sb, "agent_name", e.agentName()).append(',');
        field(sb, "tenant_id", e.tenantId()).append(',');
        field(sb, "kind", e.kind()).append(',');
        field(sb, "detail", redact.apply(e.detail())).append(',');
        nullableField(sb, "correlation_id", e.correlationId()).append(',');
        field(sb, "recorded_at", e.recordedAt().toString());
        sb.append('}');
        append(sb.toString());
    }

    /** Raw JSONL lines, for inspection. */
    public List<String> lines() {
        try {
            return Files.exists(path) ? Files.readAllLines(path, StandardCharsets.UTF_8) : List.of();
        } catch (IOException e) {
            throw new UncheckedIOException("cannot read audit log", e);
        }
    }

    public Path path() {
        return path;
    }

    // -- internals --------------------------------------------------------
    private void append(String line) {
        synchronized (lock) {
            try {
                Files.writeString(path, line + "\n", StandardCharsets.UTF_8,
                        StandardOpenOption.CREATE, StandardOpenOption.APPEND);
            } catch (IOException e) {
                throw new UncheckedIOException("cannot append to audit log", e);
            }
        }
    }

    private static StringBuilder field(StringBuilder sb, String key, String value) {
        return sb.append('"').append(key).append("\":\"").append(escape(value)).append('"');
    }

    private static StringBuilder nullableField(StringBuilder sb, String key, String value) {
        if (value == null) {
            return sb.append('"').append(key).append("\":null");
        }
        return field(sb, key, value);
    }

    private static StringBuilder raw(StringBuilder sb, String key, String jsonValue) {
        return sb.append('"').append(key).append("\":").append(jsonValue);
    }

    private static String escape(String s) {
        StringBuilder out = new StringBuilder(s.length() + 8);
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                default -> {
                    if (c < 0x20) {
                        out.append(String.format("\\u%04x", (int) c));
                    } else {
                        out.append(c);
                    }
                }
            }
        }
        return out.toString();
    }
}

File: bindings/python/src/halo_agent_harness/adapters/file_audit.py
"""Durable, append-only file AuditPort (spec §7.3).

Writes one JSON object per line (JSONL) to a file, in append mode. There is **no** update or delete
API — the log is append-only by construction (INV-4). PII is redacted before every write. This is a
minimal, dependency-free durable adapter; a JDBC / object-store adapter follows the same contract.
"""

from __future__ import annotations

import json
import threading
from datetime import datetime
from pathlib import Path
from typing import Callable

from ..ports.governance import AuditEntry, SecurityEvent
from .inmemory import redact

_DECISION = "decision"
_SECURITY = "security_event"


class FileAudit:
    """Append-only JSONL audit log. Implements ``AuditPort``; adds read-back helpers for inspection."""

    def __init__(self, path: str | Path, *, redactor: Callable[[str], str] = redact) -> None:
        self._path = Path(path)
        self._redact = redactor
        self._lock = threading.Lock()
        # Ensure the parent directory exists; create the file lazily on first append.
        self._path.parent.mkdir(parents=True, exist_ok=True)

    # -- AuditPort ------------------------------------------------------
    def record(self, entry: AuditEntry) -> None:
        self._append(
            {
                "type": _DECISION,
                "agent_name": entry.agent_name,
                "tenant_id": entry.tenant_id,
                "action": entry.action,
                "confidence": entry.confidence,
                "auto_enforced": entry.auto_enforced,
                "rationale": self._redact(entry.rationale),
                "outcome": entry.outcome,
                "correlation_id": entry.correlation_id,
                "recorded_at": entry.recorded_at.isoformat(),
            }
        )

    def record_security_event(self, event: SecurityEvent) -> None:
        self._append(
            {
                "type": _SECURITY,
                "agent_name": event.agent_name,
                "tenant_id": event.tenant_id,
                "kind": event.kind,
                "detail": self._redact(event.detail),
                "correlation_id": event.correlation_id,
                "recorded_at": event.recorded_at.isoformat(),
            }
        )

    # -- inspection (read-back) -----------------------------------------
    def entries(self) -> tuple[AuditEntry, ...]:
        return tuple(
            AuditEntry(
                agent_name=r["agent_name"],
                tenant_id=r["tenant_id"],
                action=r["action"],
                confidence=r["confidence"],
                auto_enforced=r["auto_enforced"],
                rationale=r["rationale"],
                outcome=r["outcome"],
                correlation_id=r["correlation_id"],
                recorded_at=datetime.fromisoformat(r["recorded_at"]),
            )
            for r in self._read()
            if r.get("type") == _DECISION
        )

    def security_events(self) -> tuple[SecurityEvent, ...]:
        return tuple(
            SecurityEvent(
                agent_name=r["agent_name"],
                tenant_id=r["tenant_id"],
                kind=r["kind"],
                detail=r["detail"],
                correlation_id=r["correlation_id"],
                recorded_at=datetime.fromisoformat(r["recorded_at"]),
            )
            for r in self._read()
            if r.get("type") == _SECURITY
        )

    @property
    def path(self) -> Path:
        return self._path

    # -- internals ------------------------------------------------------
    def _append(self, record: dict) -> None:
        line = json.dumps(record, ensure_ascii=False, separators=(",", ":"))
        with self._lock:
            with self._path.open("a", encoding="utf-8") as fh:
                fh.write(line + "\n")

    def _read(self) -> list[dict]:
        if not self._path.exists():
            return []
        with self._path.open("r", encoding="utf-8") as fh:
            return [json.loads(line) for line in fh if line.strip()]
